src/ParallelAlgorithms/ApparentHorizonFinder/InterpolateVolumeVars.cpp

```diff
@@ -0,0 +1,127 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "ParallelAlgorithms/ApparentHorizonFinder/InterpolateVolumeVars.hpp"
+
+#include <cstddef>
+#include <optional>
+#include <unordered_map>
+#include <vector>
+
+#include "DataStructures/Variables.hpp"
+#include "Domain/BlockLogicalCoordinates.hpp"
+#include "Domain/Domain.hpp"
+#include "Domain/ElementLogicalCoordinates.hpp"
+#include "Domain/FunctionsOfTime/FunctionOfTime.hpp"
+#include "Domain/Structure/ElementId.hpp"
+#include "NumericalAlgorithms/Interpolation/IrregularInterpolant.hpp"
+#include "ParallelAlgorithms/ApparentHorizonFinder/ComputeVarsToInterpolateToTarget.hpp"
+#include "ParallelAlgorithms/ApparentHorizonFinder/HorizonAliases.hpp"
+#include "ParallelAlgorithms/ApparentHorizonFinder/Storage.hpp"
+#include "Utilities/ErrorHandling/Assert.hpp"
+#include "Utilities/ErrorHandling/FloatingPointExceptions.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace ah {
+template <typename Fr>
+void interpolate_volume_data(
+    const gsl::not_null<ah::Storage::Iteration<Fr>*> current_iteration_storage,
+    const gsl::not_null<
+        std::unordered_map<ElementId<3>, ah::Storage::VolumeVariables<Fr>>*>
+        all_volume_variables,
+    const LinkedMessageId<double>& time, const Domain<3>& domain,
+    const domain::FunctionsOfTimeMap& functions_of_time) {
+  std::vector<ElementId<3>> element_ids;
+  element_ids.reserve(all_volume_variables->size());
+  for (const auto& [element_id, volume_vars] : *all_volume_variables) {
+    (void)volume_vars;  // Avoid clang-tidy warning
+    if (not current_iteration_storage->interpolation_is_done_for_these_elements
+                .contains(element_id)) {
+      element_ids.push_back(element_id);
+      current_iteration_storage->interpolation_is_done_for_these_elements
+          .insert(element_id);
+    }
+  }
+
+  ASSERT(current_iteration_storage->block_coord_holders.has_value(),
+         "Block logical coordinates of horizon have not been set!");
+
+  const auto& block_coord_holders =
+      current_iteration_storage->block_coord_holders.value();
+  const auto element_coord_holders =
+      element_logical_coordinates(element_ids, block_coord_holders);
+
+  for (const auto& [element_id, element_coord_holder] : element_coord_holders) {
+    const auto& offsets = element_coord_holder.offsets;
+    auto& volume_vars_storage = all_volume_variables->at(element_id);
+    auto& interpolated_vars = current_iteration_storage->interpolated_vars;
+
+    // Only fill once
+    const size_t expected_num_points = block_coord_holders.size();
+    if (interpolated_vars.number_of_grid_points() != expected_num_points) {
+      // Temporarily disable FPEs
+      const ScopedFpeState fpe_state{false};
+      interpolated_vars.initialize(
+          expected_num_points, std::numeric_limits<double>::signaling_NaN());
```

Comment:
Are the nans being tested for somewhere else?  I.e., is there a reason to override the default initialization behavior in non-debug builds?